controllers/infrastructure/byocluster_controller_test.go

```diff
@@ -0,0 +1,135 @@
+// Copyright 2021 VMware, Inc. All Rights Reserved.
+// SPDX-License-Identifier: Apache-2.0
+
+package controllers_test
+
+import (
+	"context"
+	"fmt"
+
+	. "github.com/onsi/ginkgo"
+	. "github.com/onsi/gomega"
+	infrastructurev1beta1 "github.com/vmware-tanzu/cluster-api-provider-bringyourownhost/apis/infrastructure/v1beta1"
+	"github.com/vmware-tanzu/cluster-api-provider-bringyourownhost/test/builder"
+	"k8s.io/apimachinery/pkg/types"
+	"k8s.io/client-go/kubernetes/scheme"
+	clusterv1 "sigs.k8s.io/cluster-api/api/v1beta1"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/controller/controllerutil"
+	"sigs.k8s.io/controller-runtime/pkg/reconcile"
+)
+
+var _ = Describe("Controllers/ByoclusterController", func() {
+
+	var (
+		ctx               context.Context
+		k8sClientUncached client.Client
+		byoCluster        *infrastructurev1beta1.ByoCluster
+		cluster           *clusterv1.Cluster
+	)
+
+	BeforeEach(func() {
+		ctx = context.Background()
+		var clientErr error
+
+		k8sClientUncached, clientErr = client.New(cfg, client.Options{Scheme: scheme.Scheme})
+		Expect(clientErr).NotTo(HaveOccurred())
+	})
+
+	It("should ignore when byocluster is not existed", func() {
+		_, err := byoClusterReconciler.Reconcile(ctx, reconcile.Request{
+			NamespacedName: types.NamespacedName{
+				Name:      "non-existent-byocluster",
+				Namespace: "non-existent-namespace"}})
+		Expect(err).NotTo(HaveOccurred())
+	})
+
+	It("should ignore When Cluster Controller is not set OwnerRef on ByoCluster", func() {
+		byoCluster = builder.ByoCluster(defaultNamespace, "byocluster-not-link-cluster").Build()
+		Expect(k8sClientUncached.Create(ctx, byoCluster)).Should(Succeed())
+		WaitForObjectsToBePopulatedInCache(byoCluster)
+
+		_, err := byoClusterReconciler.Reconcile(ctx, reconcile.Request{
+			NamespacedName: types.NamespacedName{
+				Name:      byoCluster.Name,
+				Namespace: byoCluster.Namespace}})
+		Expect(err).NotTo(HaveOccurred())
+	})
+
+	It("should paused when byocluster is paused", func() {
+		cluster = builder.Cluster(defaultNamespace, "cluster-paused").
+			WithPausedField(true).
+			Build()
+		Expect(k8sClientUncached.Create(ctx, cluster)).Should(Succeed())
+		WaitForObjectsToBePopulatedInCache(cluster)
+
+		byoCluster = builder.ByoCluster(defaultNamespace, "byocluster-paused").
+			WithOwnerCluster(cluster).
+			Build()
+		Expect(k8sClientUncached.Create(ctx, byoCluster)).Should(Succeed())
+
```

Comment:
Question: Do we need to add `WaitForObjectsToBePopulatedInCache(byoCluster)` here?